buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/DistributionTest.kt

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.gradlebuild.test.integrationtests
+
+import org.gradle.api.Named
+import org.gradle.api.Project
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.file.ProjectLayout
+import org.gradle.api.tasks.*
+import org.gradle.api.tasks.testing.Test
+import org.gradle.build.GradleDistribution
+import org.gradle.internal.os.OperatingSystem
+import org.gradle.process.CommandLineArgumentProvider
+import java.util.concurrent.Callable
+
+/**
+ * Base class for all tests that check the end-to-end behavior of a Gradle distribution.
+ */
+open class DistributionTest : Test() {
+
+    @get: Input
+    val operatingSystem: String by lazy {
+        val current = OperatingSystem.current()
+        // the version currently differs between our dev infrastructure, so we only track the name and the architecture
+        current.getName() + " " + System.getProperty("os.arch")
+    }
+
+    @Internal
+    val gradleInstallationForTest: GradleInstallationForTestEnvironmentProvider = GradleInstallationForTestEnvironmentProvider(project)
+
+    @Internal
+    val binaryDistributions: BinaryDistributions = BinaryDistributions(project.layout)
+
+    @Internal
+    val libsRepository: LibsRepositoryEnvironmentProvider = LibsRepositoryEnvironmentProvider(project.layout)
+
+    init {
+        dependsOn(Callable { if (binaryDistributions.distributionsRequired) listOf("all", "bin", "src").map { ":distributions:${it}Zip" } else null })
+        dependsOn(Callable { if (binaryDistributions.binZipRequired) ":distributions:binZip" else null })
+        dependsOn(Callable { if (libsRepository.required) ":toolingApi:publishLocalArchives" else null })
+        jvmArgumentProviders.add(gradleInstallationForTest)
+        jvmArgumentProviders.add(BinaryDistributionsEnvironmentProvider(binaryDistributions))
+        jvmArgumentProviders.add(libsRepository)
+    }
+}
+
+fun <K, V> Map<K, V>.asSystemPropertyJvmArguments(): Iterable<String> = this.map { (key, value) -> "-D$key=$value" }
```

Comment:
:x: can be `private` or `internal`
:x: `this` can be omitted

Reply:
It is used by `IntegrationTest`.

Reply:
yep, then `internal` should be fine

Reply:
Done.